File: src/minima_harness/tools/web_fetch.py
from __future__ import annotations

from pydantic import BaseModel, Field

from minima_harness.agent.tools import AgentTool, ToolResult, ToolUpdate, error_result
from minima_harness.ai.types import TextContent
from minima_harness.tools._exa import ExaError, exa_contents


class WebFetchParams(BaseModel):
    url: str = Field(description="The URL to fetch and read.")
    max_chars: int = Field(
        default=8000,
        ge=500,
        le=50_000,
        description="Maximum characters of page text to return (output is truncated past this).",
    )


async def _execute(
    tool_call_id: str,
    params,  # noqa: ANN001
    signal,  # noqa: ANN001
    on_update: ToolUpdate | None,
) -> ToolResult:
    assert isinstance(params, WebFetchParams)
    try:
        data = await exa_contents([params.url], max_chars=params.max_chars)
    except ExaError as exc:
        return error_result(f"web_fetch failed: {exc}")

    if not data.results:
        return error_result(f"web_fetch: no content returned for {params.url}")

    r = data.results[0]
    text = (r.text or "").strip()
    if not text:
        return error_result(f"web_fetch: page had no extractable text ({params.url})")

    suffix = ""
    if len(text) > params.max_chars:
        extra = len(text) - params.max_chars
        text = text[: params.max_chars]
        suffix = f"\n\n[truncated — {extra} more chars]"

    header = f"# {r.title}\n{params.url}\n\n" if r.title else f"{params.url}\n\n"
    body = header + text + suffix
    return ToolResult(
        content=[TextContent(text=body)],
        details={"url": params.url, "chars": len(text), "truncated": bool(suffix)},
    )


def web_fetch_tool() -> AgentTool:
    return AgentTool(
        name="web_fetch",
        description=(
            "Fetch a single URL and return its main readable text (not raw HTML). "
            "Use after web_search to read a result, or on any URL you already have. "
            "Long pages are truncated; raise max_chars if you need more."
        ),
        parameters=WebFetchParams,
        execute=_execute,
    )

File: packages/tui/src/tools/web_fetch.ts
/**
 * web_fetch — fetch a single URL's main readable text.
 * Port of minima_harness/tools/web_fetch.py.
 *
 * Uses Exa's /contents extraction when `EXA_API_KEY` is set, falling back to a keyless
 * raw fetch + HTML→text strip when the key is absent or the Exa call fails (see
 * _search.ts). The fallback's extraction is coarser than Exa's readability.
 */

import { type AgentTool, type ToolResult, errorResult } from "../agent/tools.ts";
import { text } from "../ai/types.ts";
import { WebSearchError, fetchWeb } from "./_search.ts";
import { objectSchema } from "./schema.ts";

const DEFAULT_MAX_CHARS = 8000;
const MIN_MAX_CHARS = 500;
const MAX_MAX_CHARS = 50_000;

const parameters = objectSchema(
  {
    url: { type: "string", description: "The URL to fetch and read." },
    max_chars: {
      type: "integer",
      description:
        "Maximum characters of page text to return (output is truncated past this). 500–50000.",
      default: DEFAULT_MAX_CHARS,
    },
  },
  ["url"],
);

async function execute(_id: string, params: Record<string, unknown>): Promise<ToolResult> {
  const url = String(params.url ?? "");
  const maxChars = Math.max(
    MIN_MAX_CHARS,
    Math.min(MAX_MAX_CHARS, (params.max_chars as number) ?? DEFAULT_MAX_CHARS),
  );

  let data: Awaited<ReturnType<typeof fetchWeb>>;
  try {
    data = await fetchWeb(url, maxChars);
  } catch (exc) {
    if (exc instanceof WebSearchError) return errorResult(`web_fetch failed: ${exc.message}`);
    throw exc;
  }

  if (!data.results.length) {
    return errorResult(`web_fetch: no content returned for ${url}`);
  }

  const r = data.results[0]!;
  let body = (r.text ?? "").trim();
  if (!body) {
    return errorResult(`web_fetch: page had no extractable text (${url})`);
  }

  let suffix = "";
  if (body.length > maxChars) {
    const extra = body.length - maxChars;
    body = body.slice(0, maxChars);
    suffix = `\n\n[truncated — ${extra} more chars]`;
  }

  const header = r.title ? `# ${r.title}\n${url}\n\n` : `${url}\n\n`;
  return {
    content: [text(header + body + suffix)],
    details: { url, chars: body.length, truncated: Boolean(suffix) },
  };
}

export function webFetchTool(): AgentTool {
  return {
    name: "web_fetch",
    description:
      "Fetch a single URL and return its main readable text (not raw HTML). " +
      "Use after web_search to read a result, or on any URL you already have. " +
      "Long pages are truncated; raise max_chars if you need more.",
    parameters,
    execute,
  };
}
